File: CS/CustomFilteringMVVM/MainViewModel.cs
using DevExpress.Mvvm;
using DevExpress.Mvvm.DataAnnotations;
using DevExpress.Mvvm.Xpf;
using System.Collections.ObjectModel;
using System.Linq;

namespace CustomFilteringMVVM {
    public class DataModel {
        public string Text { get; set; }
        public int Number { get; set; }
    }

    public class MainViewModel : ViewModelBase {
        public ObservableCollection<DataModel> Data { get; set; }
        public ObservableCollection<DataModel> ExcludedData { get; set; }

        public DataModel CurrentVisibleItem { get; set; }
        public DataModel CurrentExcludedItem { get; set; }

        public MainViewModel() {
            Data = new ObservableCollection<DataModel>(Enumerable.Range(0, 10).Select(i => new DataModel() { Text = $"Row {i}", Number = i }));
            ExcludedData = new ObservableCollection<DataModel>();
        }

        [Command]
        public void AddExclusion() => ExcludedData.Add(CurrentVisibleItem);

        public bool CanAddExclusion() => CurrentVisibleItem != null && !ExcludedData.Contains(CurrentVisibleItem);

        [Command]
        public void RemoveExclusion() => ExcludedData.Remove(CurrentExcludedItem);

        public bool CanRemoveExclusion() => CurrentExcludedItem != null && ExcludedData.Contains(CurrentExcludedItem);

        [Command]
        public void FilterExclusions(RowFilterArgs args) {
            if(ExcludedData.Contains(Data[args.SourceIndex])) {
                args.Visible = false;
            }
        }
    }
}

File: VB/CustomFilteringMVVM/MainViewModel.vb
Imports DevExpress.Mvvm
Imports DevExpress.Mvvm.DataAnnotations
Imports DevExpress.Mvvm.Xpf
Imports System.Collections.ObjectModel
Imports System.Linq

Namespace CustomFilteringMVVM

    Public Class DataModel

        Public Property Text As String

        Public Property Number As Integer
    End Class

    Public Class MainViewModel
        Inherits ViewModelBase

        Public Property Data As ObservableCollection(Of DataModel)

        Public Property ExcludedData As ObservableCollection(Of DataModel)

        Public Property CurrentVisibleItem As DataModel

        Public Property CurrentExcludedItem As DataModel

        Public Sub New()
            Data = New ObservableCollection(Of DataModel)(Enumerable.Range(0, 10).[Select](Function(i) New DataModel() With {.Text = $"Row {i}", .Number = i}))
            ExcludedData = New ObservableCollection(Of DataModel)()
        End Sub

        <Command>
        Public Sub AddExclusion()
            ExcludedData.Add(CurrentVisibleItem)
        End Sub

        Public Function CanAddExclusion() As Boolean
            Return CurrentVisibleItem IsNot Nothing AndAlso Not ExcludedData.Contains(CurrentVisibleItem)
        End Function

        <Command>
        Public Sub RemoveExclusion()
            ExcludedData.Remove(CurrentExcludedItem)
        End Sub

        Public Function CanRemoveExclusion() As Boolean
            Return CurrentExcludedItem IsNot Nothing AndAlso ExcludedData.Contains(CurrentExcludedItem)
        End Function

        <Command>
        Public Sub FilterExclusions(ByVal args As RowFilterArgs)
            If ExcludedData.Contains(Data(args.SourceIndex)) Then
                args.Visible = False
            End If
        End Sub
    End Class
End Namespace
